libcomposefs/lcfs-writer-erofs.c

```diff
@@ -1303,3 +1304,512 @@ int lcfs_write_erofs_to(struct lcfs_ctx_s *ctx)
 
 	return 0;
 }
+
+struct hasher_node_s {
+	uint64_t nid;
+	struct lcfs_node_s *node;
+};
+
+struct lcfs_image_data {
+	const uint8_t *erofs_data;
+	size_t erofs_data_size;
+	const uint8_t *erofs_metadata;
+	const uint8_t *erofs_metadata_end;
+	const uint8_t *erofs_xattrdata;
+	const uint8_t *erofs_xattrdata_end;
+	uint64_t erofs_build_time;
+	uint32_t erofs_build_time_nsec;
+	Hash_table *node_hash;
+};
+
+typedef union {
+	__le16 i_format;
+	struct erofs_inode_compact compact;
+	struct erofs_inode_extended extended;
+} erofs_inode;
+
+static const erofs_inode *lcfs_image_get_erofs_inode(struct lcfs_image_data *data,
+						     uint64_t nid)
+{
+	const uint8_t *inode_data = data->erofs_metadata + (nid << EROFS_ISLOTBITS);
+
+	if (inode_data >= data->erofs_metadata_end)
+		return NULL;
+
+	return (const erofs_inode *)inode_data;
+}
+
+static uint16_t erofs_inode_version(const erofs_inode *cino)
+{
+	uint16_t i_format = lcfs_u16_from_file(cino->i_format);
+	return (i_format >> EROFS_I_VERSION_BIT) & EROFS_I_VERSION_MASK;
+}
+
+static bool erofs_inode_is_compact(const erofs_inode *cino)
+{
+	return erofs_inode_version(cino) == 0;
+}
+
+static uint16_t erofs_inode_datalayout(const erofs_inode *cino)
+{
+	uint16_t i_format = lcfs_u16_from_file(cino->i_format);
+	return (i_format >> EROFS_I_DATALAYOUT_BIT) & EROFS_I_DATALAYOUT_MASK;
+}
+
+static bool erofs_inode_is_tailpacked(const erofs_inode *cino)
+{
+	return erofs_inode_datalayout(cino) == EROFS_INODE_FLAT_INLINE;
+}
+
+static struct lcfs_node_s *lcfs_build_node_from_image(struct lcfs_image_data *data,
+						      uint64_t nid);
+
+static int erofs_readdir_block(struct lcfs_image_data *data,
```

Comment:
It looks like this is a copy-paste-edit of the relevant code from cfs-fuse.c?  Might be nice to at least add a comment indicating that if one is changed, to change the other.

Or, to at least figure out how we can share more code.

Reply:
They are note quite the same because the fuse one is driven by the fuse callback apis, but yeah, we should try to share a bit more.